### DISCUSSION TITLE
Unable to launch pyppeteer within Airflow DAG (docker-compose)

### DISCUSSION THREAD

**Discussion by omairnabiel** [apache/airflow#31047] [Q&A]
I'm using the official docker-compose file provided by Airflow to spin up the container. To install the python dependencies I have a separate docker file from which the image is built and then that image is used within the docker-compose file. The Airflow spins up fine but when I try to run my task which uses pyppeteer, it fails at `launch` with the following error 

```
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/operators/python.py", line 181, in execute
    return_value = self.execute_callable()
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/operators/python.py", line 198, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/opt/airflow/dags/junaid_jamshed/scrapper.py", line 72, in scrape_website_task
    return asyncio.run(scrape_website())
  File "/usr/local/lib/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "/opt/airflow/dags/junaid_jamshed/scrapper.py", line 49, in scrape_website
    html = await get_html(women_un_stitched_url)
  File "/opt/airflow/dags/junaid_jamshed/scrapper.py", line 27, in get_html
    browser = await launch(headless=True, use_xvfb=True)
  File "/home/airflow/.local/lib/python3.7/site-packages/pyppeteer/launcher.py", line 307, in launch
    return await Launcher(options, **kwargs).launch()
  File "/home/airflow/.local/lib/python3.7/site-packages/pyppeteer/launcher.py", line 168, in launch
    self.browserWSEndpoint = get_ws_endpoint(self.url)
  File "/home/airflow/.local/lib/python3.7/site-packages/pyppeteer/launcher.py", line 227, in get_ws_endpoint
    raise BrowserError('Browser closed unexpectedly:\n')
pyppeteer.errors.BrowserError: Browser closed unexpectedly:
```

**Comment by potiuk** [comment]
NO idea what pappeteer is but Likely it tries to launch some kind of browser and it has been closed. You shoul raise it to pappeteer - my guess is that it tries to run browser interactively from inside docker to show it to the user. But it is not going to work, Airflow runs tasks in non-interactive mode on the workers run by the server and docker additionally provide isolation so that you CAN't do anything in the host (and it seems you want to run browser in the host interactively which breaks both assumptions - of airflow workers being non-interactive and docker isolating you from the host)

You need check with pappeteer if they support running whatever you want to run in non-interactive more and how it should be done - it has nothing to do with Airflow.

**Comment by omairnabiel** [comment]
So turns out that pyppeteer requires chromium drivers to be installed in order for it to work, but those don't come packaged with `apache/airflow:2.6.0`. Trying to install them using the following commands still doesn't work using apache airflow as base image. The error stating: `E: Unable to locate package google-chrome-stable`. It does work if I use some other image as base, but I want it to work with apache airflow base image

```
RUN apt-get update && apt-get install -y --no-install-recommends \
    gnupg \
    wget \
    ca-certificates \
    && wget -q -O - https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - \
    && echo "deb [arch=amd64] http://dl.google.com/linux/chrome/deb/ stable main" >> /etc/apt/sources.list.d/google.list \
    && apt-get update \
    && apt-get install -y --no-install-recommends \
    google-chrome-stable \
    && rm -rf /var/lib/apt/lists/*
    ```

**Reply by potiuk** [reply]
Then you haave to figure out how to install it on debian (this is the base image we use). you can likely find some guides by searching  - it's impossible to provide guidelines on how to install any 3rd-party software in the image - you need to figure it out yourself

**Reply by chang-1** [reply]
Were you able to install and run pyppeteer?
